contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,209 @@
+        ResourceSelectorOpGt           ResourceSelectorOperator = "Gt"
+        ResourceSelectorOpLt           ResourceSelectorOperator = "Lt"
+)
+```
+### ResourceClassStatus
+```golang
+type ResourceClassStatus struct {
+        Allocatable resources.Quantity
+        Request     resources.Quantity
+}
+```
+ResourceClass status is updated by the scheduler at:
+1. New *Resource Class* object creation.
+2. Node addition to the cluster.
+3. Node removal from the cluster.
+4. Pod creation if pod requests a resource class.
+5. Pod deletion if pod was consuming resource class.
+
+`ResourceClassStatus` serves the following two purposes:
+* Scheduler predicates evaluation while pod creation. For details, please refer further sections
+* User can view the current usage/availability details about the resource class using kubectl.
+
+### User story
+Admin knows what all devices are present on nodes and have deployed corresponding device plugins. Device plugins will make devices appear in node status. Next admin creates resource classes that have generic/portable names and metadata which can select available devices.
+1. A user submits a pod spec requesting 'X' resource classes.
+2. The scheduler filters the nodes which do not match the resource requests.
+3. scheduler selects a device for each resource class requested and annotates the pod object with device selection info.
+4. Kubelet reads the device request from pod annotation and calls `Allocate` on the matching Device Plugins.
+5. The user deletes the pod or the pod terminates
+6. Kubelet reads pod object annotation for devices consumed and calls `Deallocate` on the matching Device Plugins
+
+The scheduler is incharge of both, selecting a node and also selecting a device for requested resource classes.
+### Reason for not preferring device selection at kubelet
+Kubelet does not maintain any cache. Therefore to know the availability of a device, will have to iterate over all the existing pods running on the node. This is already done today while running predicates for each new incoming pod at kubelet. Even if we assume that scheduler cache and consumption state that is created at runtime for each pod, are exactly same, current api interfaces does not allow to pass selected device to container runtime (where actually device plugin will be invoked). This problem occurs because devices are determined internally from resource classes while other resource requests can be determined from pod object directly. To summarize, device selection at the kubelet can be done in one of the following two ways:
+* Select device at pod admission while applying predicates and change all api interfaces that are required to pass selected device to container runtime manager.
+* Create resource consumption state again at container runtime manager and select device.
+
+None of the above approach seems cleaner than doing device selection at scheduler.
```

Comment:
Is it worth mentioning here that this decision helps to retain a clean abstraction between the container runtime and kubernetes?  ISTR that was a portion of the discussion.